static/src/javascripts/projects/commercial/modules/dfp/Advert.ts

```diff
@@ -119,61 +92,35 @@ class Advert {
 
 		this.slot = slotDefinition.slot;
 		this.whenSlotReady = slotDefinition.slotReady;
-
-		this.whenLoaded = new Promise((resolve: Resolver) => {
-			this.whenLoadedResolver = resolve;
-		}).then((isLoaded: boolean): boolean => {
-			this.isLoaded = isLoaded;
-			return isLoaded;
-		});
-
-		this.whenRendered = new Promise((resolve: Resolver) => {
-			this.whenRenderedResolver = resolve;
-		}).then((isRendered) => {
-			this.isRendered = isRendered;
-			return isRendered;
-		});
-	}
-
-	static filterClasses = (
-		oldClasses: string[],
-		newClasses: string[],
-	): string[] =>
-		oldClasses.filter((oldClass) => !newClasses.includes(oldClass));
-
-	startLoading(): void {
-		this.isLoading = true;
-		this.timings.startLoading = window.performance.now();
 	}
 
-	stopLoading(isLoaded: boolean): void {
-		this.isLoading = false;
-		if (this.whenLoadedResolver) {
-			this.whenLoadedResolver(isLoaded);
-		}
-		this.timings.stopLoading = window.performance.now();
+	/**
+	 * Call this method once the ad has been rendered, it will set the
+	 * `isRendered` flag to true, which is used to determine whether to load
+	 * or refresh the ad
+	 *
+	 * @param isRendered was an advert rendered
+	 */
+	finishedRendering(isRendered: boolean): void {
+		this.isRendered = isRendered;
 	}
 
-	startRendering(): void {
-		this.isRendering = true;
-		this.timings.startRendering = window.performance.now();
-	}
+	/**
+	 * Update the "extra" classes for this slot e.g. `ad-slot--outstream`, so that the base classes
+	 * like `ad-slot` etc. are never removed
+	 *
+	 * @param newClasses An array of classes to set on the slot
+	 **/
+	async updateExtraSlotClasses(...newClasses: string[]): Promise<void> {
+		const classesToRemove = this.extraNodeClasses.filter(
+			(c) => !newClasses.includes(c),
+		);
```

Comment:
This might possibly be done in a single pass with [`element.classList.toggle`](https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/toggle) with the force flag?

Support seems very good for it.